src/CircularBuffer.mo

```diff
@@ -1,74 +1,74 @@
 import Array "mo:base/Array";
 import Nat "mo:base/Nat";
 import Iter "mo:base/Iter";
-import Debug "mo:base/Debug";
-import Int "mo:base/Int";
 
-module CircularBuffer {
-  /// Circular buffer, which preserves amount of pushed values
+import { Option = { guard }; Num = { Mod } } "Prelude";
+import { Num } "Prelude";
+
+module {
   public class CircularBuffer<T>(capacity : Nat) {
-    assert capacity != 0;
-    
     var array : [var ?T] = Array.init(capacity, null);
-    var last : Nat = 0;
-    var pushes : Int = 0;
+    var pushes : Nat = 0;
 
-    /// Number of items that were ever pushed to the buffer
-    public func pushesAmount() : Nat = Int.abs(pushes);
+    type Index = Num.Mod;
+    let { mod; add } = Mod(capacity);
 
-    /// Insert value into the buffer
-    public func push(item : T) {
-      array[last] := ?item;
-      pushes += 1;
-      last += 1;
-      if (last == capacity) last := 0;
-    };
+    public func least() : ?Index = if (pushes <= capacity) ?0 else mod pushes;
 
-    /// Return interval `[start, end)` of indices of elements available.
-    public func available() : (Nat, Nat) {
-      (Int.abs(Int.max(0, pushes - capacity)), Int.abs(pushes));
+    public func compare(x : Index, y : Index) : { #less; #equal; #greater } {
+      let ?least_ = least() else return #equal;
+
+      if (
+        x < y and (least_ <= x or least_ > y) or
+          x > y and least_ <= x and least_ > y
+      ) #less else if (x == y) #equal else #greater
     };
 
-    /// Returns single element added with number `index` or null if element is not available or index out of bounds.
-    public func get(index : Nat) : ?T {
-      let (l, r) = available();
-      if (l <= index and index < r) { array[index % capacity] } else { null };
+
+    type Id = Nat;
+
+    public func first() : ?Id = if (pushes <= capacity)
+      do ? { guard(pushes >= 1)!; 0 } else ?(pushes - capacity);
+
+    public func last() : ?Id = do ? { guard(pushes >= 1)!; pushes - 1 : Nat };
+
+    public func indexOf(id : Id) : ?Index = do ? {
+      guard(
+        id < pushes and (pushes <= capacity or id >= (pushes - capacity : Nat))
+      )!;
+      mod(id)!;
     };
 
-    /// Return iterator to values added with numbers in interval `[from; to)`.
-    /// `from` should be not more then `to`. Both should be not more then `pushes`.
-    public func slice(from : Nat, to : Nat) : Iter.Iter<T> {
-      assert from <= to;
-      let interval = available();
 
-      assert interval.0 <= from and from <= interval.1 and interval.0 <= to and to <= interval.1;
+    public func pushesAmount() : Nat = pushes;
+
+    public func push(item : T) = ignore do ? {
+      array[mod(pushes)!] := ?item;
```

Comment:
can we make mod to never return null? For instance we can return x % 1, which is as useless as x % 0, but does not produce error. Those `!` everywhere are potential source of bugs

Reply:
It will return `null` only for `% 0` capacity. Looks safe to me, because type system checks it.

Reply:
Keeps `assert`s in code that letting some data inputs break our code feels worse.